File: ios/Packages/FatwaBotKit/Sources/CoreKit/SemVer.swift
import Foundation

/// Dotted-numeric version comparison for rollout gates ("1.2.10" ≥ "1.2.9").
/// Non-numeric fragments compare as 0; missing components as 0.
public enum SemVer {
    public static func isVersion(_ version: String, atLeast minimum: String) -> Bool {
        let lhs = components(version)
        let rhs = components(minimum)
        for index in 0..<max(lhs.count, rhs.count) {
            let a = index < lhs.count ? lhs[index] : 0
            let b = index < rhs.count ? rhs[index] : 0
            if a != b { return a > b }
        }
        return true
    }

    private static func components(_ version: String) -> [Int] {
        version.split(separator: ".").map { Int($0.prefix(while: \.isNumber)) ?? 0 }
    }
}

File: android/core/common/src/main/kotlin/com/fatwabot/core/common/SemVer.kt
package com.fatwabot.core.common

/**
 * Dotted-numeric version comparison for rollout gates ("1.2.10" ≥ "1.2.9").
 * Mirror of iOS CoreKit.SemVer. Non-numeric fragments compare as 0.
 */
object SemVer {
    fun isVersionAtLeast(version: String, minimum: String): Boolean {
        val lhs = components(version)
        val rhs = components(minimum)
        for (i in 0 until maxOf(lhs.size, rhs.size)) {
            val a = lhs.getOrElse(i) { 0 }
            val b = rhs.getOrElse(i) { 0 }
            if (a != b) return a > b
        }
        return true
    }

    private fun components(version: String): List<Int> =
        version.split(".").map { part -> part.takeWhile { it.isDigit() }.toIntOrNull() ?: 0 }
}
